web/modules/custom/joinup_core/joinup_core.post_update.php

```diff
@@ -681,3 +681,104 @@ function joinup_core_post_update_install_publication_date() {
+  // if the entity has never been published or the publication date if the
+  // entity has been published at least once, regardless of the current state.
+  $database = \Drupal::database();
+  $node_storage = \Drupal::entityTypeManager()->getStorage('node');
+  if (empty($sandbox['nids'])) {
+    // Clean all set publication dates from the revisions table, as the empty
+    // value is needed later on.
+    $database
+      ->update($node_storage->getRevisionDataTable())
+      ->expression('published_at', 'null')
+      ->execute();
+
+    $sandbox['nids'] = $database->query('SELECT nid FROM {node} n')->fetchCol();
+    $sandbox['progress'] = 0;
+    $sandbox['total'] = count($sandbox['nids']);
+  }
+
+  $nids = array_splice($sandbox['nids'], 0, 300);
+  $query = $database->select('node', 'n');
+  $query->leftJoin('node_field_revision', 'nr', 'n.nid = nr.nid');
+  $query->fields('nr', ['nid', 'vid', 'changed']);
+  $query->condition('status', 1);
+  $query->condition('n.nid', $nids, 'IN');
+  $query->groupBy('nr.nid, nr.vid, nr.changed');
+  $query->orderBy('nr.vid', 'DESC');
+
+  $node_data = $query->execute()->fetchAllAssoc('nid');
+  foreach ($nids as $nid) {
+    if (empty($node_data[$nid])) {
+      // Node has never been published. Set the default value to all revisions.
+      $database
+        ->update($node_storage->getDataTable())
+        ->expression('published_at', PUBLICATION_DATE_DEFAULT)
+        ->condition('nid', $nid)
+        ->execute();
+      $database
+        ->update($node_storage->getRevisionDataTable())
+        ->expression('published_at', PUBLICATION_DATE_DEFAULT)
+        ->condition('nid', $nid)
+        ->execute();
+    }
+    else {
+      // Node has been published at least once. The data table will receive the
+      // publication date while for the revision table, the publication date
+      // will be set for all versions including and following the first
+      // published version. All others will receive the default value.
+      $first_published_timestamp = $node_data[$nid]->changed;
+      $first_published_vid = $node_data[$nid]->vid;
+
+      // Set the revisions that will receive the publication timestamp.
+      $database
+        ->update($node_storage->getDataTable())
+        ->expression('published_at', $first_published_timestamp)
+        ->condition('nid', $nid)
+        ->execute();
+      $database
+        ->update($node_storage->getRevisionDataTable())
+        ->expression('published_at', $first_published_timestamp)
+        ->condition('nid', $nid)
+        ->condition('vid', $first_published_vid, '>=')
+        ->execute();
+
+      // The revisions that are left have a null publication date and need to
+      // receive the default value.
+      $database
+        ->update($node_storage->getRevisionDataTable())
+        ->expression('published_at', PUBLICATION_DATE_DEFAULT)
+        ->condition('nid', $nid)
+        ->isNull('published_at')
+        ->execute();
```

Comment:
This last step can also be done in a single query, at the end of the entire batch operation. If we have 20000 published nodes, this now does 20000 queries. But if we wait until all the published revisions are processed, and then run this query at the end (without the `->condition('nid', $nid)` call) then we only need to do 1 query.